### PR TITLE
[SPARK-49656][SS] Add support for state variables with value state collection types and read change feed options

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add support for state variables with value state collection types and read change feed options

### Why are the changes needed?
Without this, we cannot support reading per key changes for state variables used with stateful processors.

### Does this PR introduce _any_ user-facing change?
Yes

Users can now query value state variables with the following query:

```
        val changeFeedDf = spark.read
            .format("statestore")
            .option(StateSourceOptions.PATH, <checkpoint_loc>)
            .option(StateSourceOptions.STATE_VAR_NAME, <state_var_name>)
            .option(StateSourceOptions.READ_CHANGE_FEED, true)
            .option(StateSourceOptions.CHANGE_START_BATCH_ID, 0)
            .load()
```

### How was this patch tested?
Added unit tests

```
[info] Run completed in 17 seconds, 318 milliseconds.
[info] Total number of tests run: 2
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 2, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @jingz-db @HeartSaVioR - PTAL, thx !

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala [hunk 1]
```diff
@@ -646,15 +646,14 @@ class RocksDB(
           // is enabled.
           if (shouldForceSnapshot.get()) {
             uploadSnapshot()
+            shouldForceSnapshot.set(false)
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala:649]: Have to add the fix temporarily till https://github.com/apache/spark/pull/48125 is merged

**Review by HeartSaVioR — COMMENTED** [review]
First pass. Mostly style and suggestion for refactor.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 2]
```diff
@@ -676,27 +681,83 @@ class RocksDBStateStoreChangeDataReader(
...
+
+      breakable {
+        while (true) {
+          val reader = currentChangelogReader()
+          if (reader == null) {
+            return null
+          }
+
+          currRecord = reader.next()
+          currEncoder = keyValueEncoderMap.get(colFamilyNameOpt
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 3]
```diff
@@ -676,27 +681,83 @@ class RocksDBStateStoreChangeDataReader(
...
+    if (colFamilyIdBytesOpt.isDefined) {
+      // If we are reading records for a particular column family, the corresponding vcf id
+      // will be encoded in the key byte array. We need to extract that and compare for the
+      // expected column family id. If it matches, we return the record. If not, we move to
+      // the next record. Note that this has be handled across multiple changelog files and we
+      // rely on the currentChangelogReader to move to the next changelog file when needed.
+      var currRecord: (RecordType.Value, Array[Byte], Array[Byte]) = null
+      var currEncoder: (RocksDBKeyStateEncoder, RocksDBValueStateEncoder) = null
+
+      breakable {
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 4]
```diff
@@ -676,27 +681,83 @@ class RocksDBStateStoreChangeDataReader(
...
+        (currRecord._1, keyRow, valueRow, currentChangelogVersion - 1)
+      }
     } else {
-      val valueRow = rocksDBValueStateEncoder.decodeValue(valueArray)
-      (recordType, keyRow, valueRow, currentChangelogVersion - 1)
+      val reader = currentChangelogReader()
+      if (reader == null) {
+        return null
+      }
+      val (recordType, keyArray, valueArray) = reader.next()
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala]: Done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 5]
```diff
@@ -676,27 +681,83 @@ class RocksDBStateStoreChangeDataReader(
...
+          }
+
+          currRecord = reader.next()
+          currEncoder = keyValueEncoderMap.get(colFamilyNameOpt
+            .getOrElse(StateStore.DEFAULT_COL_FAMILY_NAME))
+
+          val colFamilyIdBytes: Array[Byte] = colFamilyIdBytesOpt.get
+
+          val keyArrayWithColFamilyId = new Array[Byte](colFamilyIdBytes.size)
+          Array.copy(currRecord._2, 0, keyArrayWithColFamilyId, 0, colFamilyIdBytes.size)
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala]: Done

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**